src/connectors/impls/gcs/writer.rs

```diff
@@ -0,0 +1,682 @@
+// limitations under the License.
+
+use crate::connectors::prelude::{
+    Attempt, ErrorKind, EventSerializer, Result, SinkAddr, SinkContext, SinkManagerBuilder,
+    SinkReply, Url,
+};
+use crate::connectors::sink::Sink;
+use crate::connectors::utils::url::HttpsDefaults;
+use crate::connectors::{
+    CodecReq, Connector, ConnectorBuilder, ConnectorConfig, ConnectorType, Context,
+};
+use crate::system::KillSwitch;
+use async_std::task::sleep;
+#[cfg(not(test))]
+use gouth::Token;
+#[cfg(not(test))]
+use http_client::h1::H1Client;
+use http_client::HttpClient;
+use http_types::{Method, Request};
+use std::time::Duration;
+use tremor_pipeline::{ConfigImpl, Event};
+use tremor_value::Value;
+use value_trait::ValueAccess;
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct Config {
+    #[serde(default = "default_endpoint")]
+    endpoint: String,
+    #[serde(default = "default_connect_timeout")]
+    #[cfg_attr(test, allow(unused))]
+    connect_timeout: u64,
+    #[serde(default = "default_buffer_size")]
+    buffer_size: usize,
+}
+
+fn default_endpoint() -> String {
+    "https://storage.googleapis.com/upload/storage/v1".to_string()
+}
+
+fn default_connect_timeout() -> u64 {
+    10_000_000_000
+}
+
+fn default_buffer_size() -> usize {
+    1024 * 1024 * 8 // 8MB - the recommended minimum
+}
+
+impl ConfigImpl for Config {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        ConnectorType("gcs_writer".into())
+    }
+
+    async fn build_cfg(
+        &self,
+        _alias: &str,
+        _config: &ConnectorConfig,
+        connector_config: &Value,
+        _kill_switch: &KillSwitch,
+    ) -> Result<Box<dyn Connector>> {
+        let config = Config::new(connector_config)?;
+
+        if config.buffer_size % (256 * 1024) != 0 {
+            return Err("Buffer size must be a multiple of 256kiB".into());
+        }
+        let url = Url::<HttpsDefaults>::parse(&config.endpoint)?;
+
+        Ok(Box::new(GCSWriterConnector { config, url }))
+    }
+}
+
+struct GCSWriterConnector {
+    config: Config,
+    url: Url<HttpsDefaults>,
+}
+
+#[async_trait::async_trait]
+impl Connector for GCSWriterConnector {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let sink = GCSWriterSink {
+            client: None,
+            url: self.url.clone(),
+            config: self.config.clone(),
+            buffers: ChunkedBuffer::new(self.config.buffer_size),
+            current_name: None,
+            current_session_url: None,
+        };
+
+        let addr = builder.spawn(sink, sink_context)?;
+        Ok(Some(addr))
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct ChunkedBuffer {
```

Comment:
This reminds me a lot of https://docs.rs/bytes/latest/bytes/struct.BytesMut.html